pkg/reconciler/service/resources/rolloutorchestrator.go

```diff
@@ -0,0 +1,202 @@
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/ptr"
+	v1 "knative.dev/serving-progressive-rollout/pkg/apis/serving/v1"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	"knative.dev/serving/pkg/apis/serving"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+var (
+	// OverSubRatio is the parameter, that determines how much percentage of the traffic to shift
+	// from the old to the new revision during each stage in the progressive rollout.
+	OverSubRatio = 10
+)
+
+// RevisionRecord is a struct that hosts the name, minScale and maxScale for the revision.
+type RevisionRecord struct {
+	MinScale *int32
+	MaxScale *int32
+	Name     string
+}
+
+// ReadIntAnnotation reads the int value of a specific key in the annotation of the revision.
+func ReadIntAnnotation(revision *servingv1.Revision, key string) (result *int32) {
+	if val, ok := revision.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntServiceAnnotation reads the int value of a specific key in the annotation of the service.
+func ReadIntServiceAnnotation(service *servingv1.Service, key string) (result *int32) {
+	if val, ok := service.Spec.Template.ObjectMeta.Annotations[key]; ok {
+		i, err := strconv.ParseInt(val, 10, 32)
+		if err == nil {
+			result = ptr.Int32(int32(i))
+		}
+	}
+	return
+}
+
+// ReadIntRevisionRecord reads the minScale and maxScale in the RevisionRecord.
+func ReadIntRevisionRecord(val RevisionRecord) (min *int32, max *int32) {
+	if val.MinScale != nil {
+		min = ptr.Int32(*val.MinScale)
+	}
+	if val.MaxScale != nil {
+		max = ptr.Int32(*val.MaxScale)
+	}
+	return min, max
+}
+
+func initializeTargetRevisions(ultimateRevisionTarget *[]v1.TargetRevision, traffic *servingv1.TrafficTarget,
+	index int, lastRevName string, service *servingv1.Service, records map[string]RevisionRecord) {
+	target := v1.TargetRevision{}
+	if traffic.RevisionName == "" || traffic.RevisionName == lastRevName {
+		target.LatestRevision = ptr.Bool(true)
+		target.RevisionName = lastRevName
+	} else {
+		target.LatestRevision = ptr.Bool(false)
+		target.RevisionName = traffic.RevisionName
+	}
+	if traffic.Percent == nil {
+		target.Percent = ptr.Int64(100)
+	} else {
+		target.Percent = ptr.Int64(*traffic.Percent)
+	}
+
+	target.ConfigurationName = traffic.ConfigurationName
+	target.Tag = traffic.Tag
+	target.URL = traffic.URL
+	if val, ok := records[target.RevisionName]; ok {
+		target.MinScale, target.MaxScale = ReadIntRevisionRecord(val)
+	} else {
+		// Get min and max scales from the service
+		target.MinScale = ReadIntServiceAnnotation(service, autoscaling.MinScaleAnnotationKey)
+		target.MaxScale = ReadIntServiceAnnotation(service, autoscaling.MaxScaleAnnotationKey)
+	}
```

Comment:
Probably not safe to initialize from service min/max scale as it could be an old revision